### ISSUE TITLE
Wrong mask preparing pipeline in llava_arch.py

### ISSUE THREAD

**Issue by Fuzm7** [orailix/PACT#4]
I believe there's an issue when preparing the 0/1 mask for the image/text, specifically on line 659 in [[llava_arch.py.]](https://github.com/orailix/PACT/blob/main/LLaVA-NeXT/llava/model/llava_arch.py)
The right way to do it is following the original pipeline in LLaVA like "attention_mask" and "position_ids", which is populating a boolean tensor with all zeros:

new_labels_padded = torch.full((batch_size, max_len), IGNORE_INDEX, dtype=new_labels[0].dtype, device=new_labels[0].device)
attention_mask = torch.zeros((batch_size, max_len), dtype=attention_mask.dtype, device=attention_mask.device)
position_ids = torch.zeros((batch_size, max_len), dtype=position_ids.dtype, device=position_ids.device)

**Comment by Mohamed-Dhouib** [comment]
Hello,
The mask you are referring to in Line 659 is not the attention mask. It is a mask indicating the position of visual tokens (as we only apply reduction on visual tokens so the mask is needed later).  
We later concatenate it to positions_ids (we do so to avoid changing the whole code by adding more variables):
https://github.com/orailix/PACT/blob/4b4b112d692f588207ba1c774cb79748a774430e/LLaVA-NeXT/llava/model/llava_arch.py#L698
and we unpack it in: 
https://github.com/orailix/PACT/blob/4b4b112d692f588207ba1c774cb79748a774430e/transformers/models/qwen2/modeling_qwen2.py#L966
https://github.com/orailix/PACT/blob/4b4b112d692f588207ba1c774cb79748a774430e/transformers/models/qwen2/modeling_qwen2.py#L995

**Comment by Fuzm7** [comment]
Right, I understand it's a boolean tensor representing images. Your main issue is directly stacking them, and the problem now is that you can't guarantee the same number of tokens within each batch. You're currently stacking them directly.

**Comment by Mohamed-Dhouib** [comment]
Token reduction is done at inference time, the batch size is always one.

**Comment by Mohamed-Dhouib** [comment]
Token reduction is done at inference time, the batch size is set to one when doing inference

**Comment by Fuzm7** [comment]
Instead of restricting the current framework to a single batch, which is more of a theoretical exercise, it would be more practical to build it using the `max_len` approach from the original paper.
For example:
  attention_mask = torch.zeros((batch_size, max_len), dtype=attention_mask.dtype, device=attention_mask.device)
  position_ids = torch.zeros((batch_size, max_len), dtype=position_ids.dtype, device=position_ids.device)

**Comment by Fuzm7** [comment]
Even the way you handled it afterward could easily lead to misunderstandings. Following the stack operation described above, readers might naturally assume the mask is a tensor with shape (bsz, length, 1). However, in reality, it's (1, length, 1) as you pointed out. Only in this case does the subsequent code work, because `squeeze` directly transforms the mask into a one-dimensional tensor, allowing it to be concatenated with `Position_ids`:
 
            position_ids = torch.arange(
                    0, new_input_embeds.shape[1], dtype=torch.long, device=new_input_embeds.device
                ).unsqueeze(1).unsqueeze(0)
            
            mask=torch.tensor(mask, dtype=torch.bool).squeeze().unsqueeze(1).unsqueeze(0)

which on line 690 to 694 at [[llava_arch.py.]](https://github.com/orailix/PACT/blob/main/LLaVA-NeXT/llava/model/llava_arch.py)

**Comment by Fuzm7** [comment]
Don't you think the current coding style is rather inelegant?

**Comment by Mohamed-Dhouib** [comment]
> Instead of restricting the current framework to a single batch, which is more of a theoretical exercise, it would be more practical to build it using the `max_len` approach from the original paper. For example: attention_mask = torch.zeros((batch_size, max_len), dtype=attention_mask.dtype, device=attention_mask.device) position_ids = torch.zeros((batch_size, max_len), dtype=position_ids.dtype, device=position_ids.device)

Yeah I agree that it would be more useful that way, however the goal of this repo is to compare different token reduction techniques, not to have a production ready model.

**Comment by Mohamed-Dhouib** [comment]
> Even the way you handled it afterward could easily lead to misunderstandings. Following the stack operation described above, readers might naturally assume the mask is a tensor with shape (bsz, length, 1). However, in reality, it's (1, length, 1) as you pointed out. Only in this case does the subsequent code work, because `squeeze` directly transforms the mask into a one-dimensional tensor, allowing it to be concatenated with `Position_ids`:
> 
> ```
>         position_ids = torch.arange(
>                 0, new_input_embeds.shape[1], dtype=torch.long, device=new_input_embeds.device
>             ).unsqueeze(1).unsqueeze(0)
>         
>         mask=torch.tensor(mask, dtype=torch.bool).squeeze().unsqueeze(1).unsqueeze(0)
> ```
> 
> which on line 690 to 694 at [[llava_arch.py.]](https://github.com/orailix/PACT/blob/main/LLaVA-NeXT/llava/model/llava_arch.py)

Thanks for the feedback, I will add comments to avoid confusion.

**Comment by Mohamed-Dhouib** [comment]
What makes the modification done to create the mask a bit complicated to understand is that we don’t want to include special tokens as visual tokens. Otherwise, creating the mask would have been straightforward. 
Feel free to ask about any part of the code that isn’t clear.

**Comment by Mohamed-Dhouib** [comment]
What makes the modification done to create the mask a bit complicated to understand is that we don’t want to include special tokens as visual tokens. Otherwise, creating the mask would have been straightforward. 
To understand the logic, keep in mind that we mirror how the embeddings are created for llava-OneVision  and progressively  create the mask following the same logic while associating one to visual tokens.
Feel free to ask about any part of the code that isn’t clear.
